### PR TITLE
Faster SequenceRecord initialization.

### PR DESCRIPTION
**Pull request by rhpvorderman**
Before:

```
Benchmark 1: python dnaio_read.py ~/test/big2.fastq
  Time (mean ± σ):     914.5 ms ±  23.8 ms    [User: 792.7 ms, System: 121.7 ms]
  Range (min … max):   882.0 ms … 950.4 ms    10 runs
```

After:
```
Benchmark 1: python dnaio_read.py ~/test/big2.fastq
  Time (mean ± σ):     752.6 ms ±  14.0 ms    [User: 628.9 ms, System: 123.5 ms]
  Range (min … max):   725.8 ms … 779.4 ms    10 runs
```

The disadvantage is that we break `__new__` for everyone. Is this a problem for cutadapt? `__init__` works fine.

The speed advantage requires some explanation. Using the CPython API there are two methods you can set on your extension class: `tp_new` and `tp_init`. `tp_new` creates the class and `tp_init` initializes it. Most extension classes have however, no dynamic attributes so do not need further initialization. tp_new also accepts arguments and keyword arguments and can therefore be used to set the basic attributes. Therefore some objects such as the python `bytes` object only use the `tp_new` slot.
Also in the Python C-API it is possible to forego the `tp_new` slot entirely and simply initialize objects using `PyObject_New` and set struct members manually.  This is not possible in Cython as Cython needs full control over how objects are initialized (it might add/need extra members to the struct, so it is not possible to preduct this beforehand).

I did some digging and this is how the tp_new function looked before:
```C
static PyObject *__pyx_tp_new_5dnaio_5_core_SequenceRecord(PyTypeObject *t, PyObject *a, PyObject *k) {
  struct __pyx_obj_5dnaio_5_core_SequenceRecord *p;
  PyObject *o;
  if (likely((t->tp_flags & Py_TPFLAGS_IS_ABSTRACT) == 0)) {
    o = (*t->tp_alloc)(t, 0);
  } else {
    o = (PyObject *) PyBaseObject_Type.tp_new(t, __pyx_empty_tuple, 0);
  }
  if (unlikely(!o)) return 0;
  p = ((struct __pyx_obj_5dnaio_5_core_SequenceRecord *)o);
  p->_name = Py_None; Py_INCREF(Py_None);
  p->_sequence = Py_None; Py_INCREF(Py_None);
  p->_qualities = Py_None; Py_INCREF(Py_None);
  if (unlikely(__pyx_pw_5dnaio_5_core_14SequenceRecord_1__cinit__(o, a, k) < 0)) goto bad;
  return o;
  bad:
  Py_DECREF(o); o = 0;
  return NULL;
}
```
As you can see, it always calls the `__cinit__` function which has uses arguments (`a`) and keyword arguments (`k`). Most of what `__cinit__` does is parsing the argument tuple and keyword dict. This is expensive. When `__cinit__` is removed this is the tp_new function.

```C
static PyObject *__pyx_tp_new_5dnaio_5_core_SequenceRecord(PyTypeObject *t, CYTHON_UNUSED PyObject *a, CYTHON_UNUSED PyObject *k) {
  struct __pyx_obj_5dnaio_5_core_SequenceRecord *p;
  PyObject *o;
  if (likely((t->tp_flags & Py_TPFLAGS_IS_ABSTRACT) == 0)) {
    o = (*t->tp_alloc)(t, 0);
  } else {
    o = (PyObject *) PyBaseObject_Type.tp_new(t, __pyx_empty_tuple, 0);
  }
  if (unlikely(!o)) return 0;
  p = ((struct __pyx_obj_5dnaio_5_core_SequenceRecord *)o);
  p->_name = Py_None; Py_INCREF(Py_None);
  p->_sequence = Py_None; Py_INCREF(Py_None);
  p->_qualities = Py_None; Py_INCREF(Py_None);
  return o;
}
```
As you can see, it adds the UNUSED macros to the arguments and keyword arguments. By default it sets everything to `None` which is a waste of resources, but not as much as parsing keyword arguments.

TLDR: I did it! Finally I, managed to make Cython do the most barebones init possible, while still letting Cython do the init and not relying on hacks trough the C-API. This should be stable regardless of any new Cython version going forward.

EDIT: I kept a branch in pure C around. It was very useful for learning some parts of the C-API and asses the maximum capability of our current approach. It performs like this:
```
Benchmark 1: python dnaio_read.py ~/test/big2.fastq
  Time (mean ± σ):     683.2 ms ±  19.7 ms    [User: 565.2 ms, System: 117.7 ms]
  Range (min … max):   653.1 ms … 710.8 ms    10 runs
```
So it is even faster (~10%) because it foregoes all the Cython stuff, but this comes at a cost of much verbose. We have already discussed this and dumping Cython which you are very familiar with, for a verbose C code base that is more difficult to maintain is not helpful. Having the speed difference down to 10% makes it definitely not worth the switch. (Still I prefer C for my own projects as getting Cython to do the things I actually want has been quite an annoying and time-consuming process ;-). But finally, that is now finished for dnaio.)

### CODE REVIEW

**Comment by rhpvorderman** [comment]
The post above is quite a ramble. Let me summarize now that I have a more clear head:
Any python object needs to be allocated in memory. This is the purpose of the `__new__` function. The `__new__` function can also be used to initialize the object if its attributes are static.
The `__init__` function is used on objects on which its attributes are dynamic and context dependent. Python automatically calls the objects `__new__` function before `__init__` is called. 

Cython implements it as follows.
- The `__new__` function is called to initialize a cdef class. In this `__new__` function all python object members are set to None.
- The `__cinit__` function is called from the `__new__` function to initialize all the C members of the object (if the user has specified `__cinit__`). `__cinit__` does not take C arguments, but an a python argument tuple and keyword dict instead. Which are parsed.
- The `__init__` function is implemented by the user. This also takes a python argument tuple and keyword dict. 

By leaving out `__cinit__` entirely, it is not called anymore. That means that an extra step parsing an argument tuple and a keyword dict is left out. This is a major source of slowdown. We don't need it because there are no C attributes that need initializing in the SequenceRecord class.

File: src/dnaio/_core.pyx [hunk 1]
```diff
@@ -85,12 +85,6 @@ cdef class SequenceRecord:
...
         object _qualities
 
-    def __cinit__(self, object name, object sequence, object qualities=None):
-        """Set qualities to None if there are no quality values"""
-        self._name = name
-        self._sequence = sequence
-        self._qualities = qualities
-
     def __init__(self, object name, object sequence, object qualities = None):
         # __cinit__ is called first and sets all the variables.
```
Comment (marcelm) [inline src/dnaio/_core.pyx:89]: This can be deleted now

**Review by marcelm — APPROVED** [review]
Cool! Thanks also for the second comment; that was helpful. Cutadapt doesn’t use `__new__`, so that not being available isn’t a problem, and I think `__init__` is indeed good enough.

Can you add a changelog entry? Then feel free to merge!

Probably time for a release (0.9.2) afterwards. I’ll look into that.

**Comment by rhpvorderman** [comment]
>  and I think __init__ is indeed good enough.

By not calling `__cinit__` (via `__new__`), `__init__` should also be faster. Unfortunately this is hard to verify with timeit. The old `__cinit__` code was pages long, mostly for parsing arguments (it hardly did anything else). I am glad we have been able to optimize it out.

**Comment by rhpvorderman** [comment]
> Probably time for a release (0.9.2) afterwards. I’ll look into that.

I propose merging #96 too and bumping to 0.10.0. Then we can use that version for the benchmarks in the paper as it should be faster.
